### DISCUSSION TITLE
Is there an interest from the maintainers in providing examples to the stdlib functions?

### DISCUSSION THREAD

**Discussion by benjamin-thomas** [ocaml/ocaml#12129] [Documentation]
Hello!

Following the discussion at: https://discuss.ocaml.org/t/why-the-ocaml-ecosystem-is-poorly-documented/11649/24, I  argued that providing examples related to functions could provide a "better" user experience.

How do the maintainers feel about such an idea?

I've read "OCaml from the very beginning" and next read for me is "More OCaml" from the same author.

So I'm probably at a good place to discover, and eventually document certain function usages (if there is any interest of course).

**Comment by xavierleroy** [comment]
As @dbuenzli mentioned in the discuss.ocaml.org thread, I would rather not have examples attached to individual functions.

For one thing, users who are already familiar with a library prefer to have short function descriptions in English or as concise algebraic specifications; long examples are just distracting.

For another thing, many library functions cannot meaningfully be described by examples involving only this function.  E.g. for the Queue module of the OCaml standard library, you cannot show what `Queue.add` or `Queue.take` does without involving the other function.

For these two reasons, I think it's better to add examples at the level of whole modules.  This way you can show the interaction of `Queue.add` and `Queue.take` and `Queue.empty` in a meaningful manner.  For reference, you could look at the `Examples` sections that were added recently to stdlib modules in pull request https://github.com/ocaml/ocaml/pull/11476 .  That's the kind of things we are interested in; not per-function examples.

**Reply by benjamin-thomas** [reply]
Thanks for your feedback.

It seems to me that certain functions can be perfectly understood in isolation. For those, I think a small example or two would help drive the point.

For the case where a function doesn’t stand on its own and must be composed, then I agree with your observation and I wouldn’t expect to find anything of value documented "inline". I would instinctively look elsewhere (but that would also break my flow).

Regarding the fact that such examples could be distracting, I’d like to push back with two arguments:

1. for beginners more doc = better. So by not providing examples, advanced users are favored over newbies, making the language potentially less welcoming/attractive to some.
2. I assume that advanced OCaml programmers know the stdlib fairly well and thus don’t actually need to inspect the documentation (the type annotations suffices)

Furthermore, I’ve just looked at this PR in more details and I see some pretty beefy function doc with examples, here:

- https://github.com/ocaml/ocaml/blob/2231b5e83c504e743e3c6c18f6bf38ebaa865b1a/stdlib/format.mli#L1262
- https://github.com/ocaml/ocaml/blob/2231b5e83c504e743e3c6c18f6bf38ebaa865b1a/stdlib/hashtbl.mli#L543

Which is what I was arguing for.

So if I understand things correctly, these two examples are actually representative of undesirable documentation?

To put things into perspective, when I engaged in this conversation (on the forum), I just saw a mostly comparitevely "dry" doc for the list module: https://github.com/ocaml/ocaml/blob/2231b5e83c504e743e3c6c18f6bf38ebaa865b1a/stdlib/list.mli

**Comment by Octachron** [comment]
The assumption that only beginners read documentation doesn't match my experience. All developers read reference documentations. 

Similarly, I am not sure where does the idea that verbosity always helps beginners comes from? Better documentation doesn't mean more documentation. 

Your emphasis on beginners made me think that it would most probably a worthwhile endeavor to have tutorials and how-to guides documentation for all the standard library modules as stated in the discussion that you linked. But a good tutorial does not make for a good reference documentation.  

For your examples, in the `Format` case we have an example showing how the `Format` primitives are mapped to the format string specifiers. That is a pretty special case of a strongly typed DSL with special syntactic support; it doesn't make much sense to try to infer a generic rule from this specific case and I wouldn't be opposed to have one such example in the `Printf` module too.

In your second example, you are misreading the documentation: there are no examples in the documentation of `seeded_hash_param`, what you read as an example is a (far too laconic) explanation of the argument of the function.

**Comment by benjamin-thomas** [comment]
> Similarly, I am not sure where does the idea that verbosity always helps beginners comes from? Better documentation doesn't mean more documentation.

For sure, but less isn't necessarily more either :)

I can only give my personal feedback exploring the language. That's where this feeling comes from. To me more is better when you don't know what you don't know.

It sounds like I am not very convincing so I will try to give one last argument. Have you heard of the Gleam language? It's a statically typed functional language that targets the BEAM vm. I've only dabbled with it but I can say the "beginner experience" is good. And I feel like they've done a great job at documentation.

Here is their doc for the list module: https://hexdocs.pm/gleam_stdlib/gleam/list.html

This is the content you see when you ask your editor to show you the function's doc.

It's as simple as can be, while very informative (look at `intersperse` for instance).

I would like to see OCaml doc like this.

But if this doesn't resonate with you, that's fine :)

I just wanted to make you aware of how things can be different.

Have a great day

**Comment by gasche** [comment]
Personally I think there would be space for per-function examples, but that we don't have the tooling in place to make this good right now:

1. Obviously we would want those examples to be checked (compiled and executed), and the stdlib doc does not currently support this. (The reference manual does.)
2. I understand the point that per-function examples add noise to the documentation; ideally they would be made unobtrusive by being hidden-by-default in the global page, or only available in per-function documentation or something like that. This is also something we don't handle well.

However, if there is resistance to the idea of per-function examples, I think there are still many ways we can improve the documentation (not just of the stdlib, of many OCaml libraries) that are more consensual: we could better structure their API documentation, provide module-scale examples, provide tutorials, etc. I wouldn't put "having per-function example" at the top of a documentation-improvement TODO list, so it is easier to do little about it for now.

Finally; let me articulate why I think that per-function examples have value. The current documentation is optimized for people who can read a precise informal description of what a function does, and immediately be comfortable using this function. This is fine for many programmers, including basically everyone that contributes to the stdlib right now, but it still leaves people out:

1. The use of informal English as the main specification language creates a language barrier. For example:

   ```ocaml
   val append : 'a list -> 'a list -> 'a list
   (** [append l0 l1] appends [l1] to [l0].
        Same function as the infix operator [@].
        @since 5.1 this function is tail-recursive.
    *)
   ```
   
   I believe that some programmers who are not comfortable with English would immediately grasp what this function does by looking at two examples, but may have trouble understanding the current documentation string if they do not already know what `append` means.

2. The use of *abstract* informal text creates a barrier to understanding for people without a strong scientific education. For example:

   ```ocaml
   val mem : 'a -> 'a array -> bool
   (** [mem a set] is true if and only if [a] is structurally equal
       to an element of [l] (i.e. there is an [x] in [l] such that
       [compare a x = 0]).
       @since 4.03 *)
   ```
   
   This documentation is precise but I believe that many beginners would need to read it several times to understand what it means.
   
   ```ocaml
   val compare : ('a -> 'a -> int) -> 'a list -> 'a list -> int
   (** [compare cmp [a1; ...; an] [b1; ...; bm]] performs
       a lexicographic comparison of the two input lists,
       using the same ['a -> 'a -> int] interface as {!Stdlib.compare}:
   
       - [a1 :: l1] is smaller than [a2 :: l2] (negative result)
         if [a1] is smaller than [a2], or if they are equal (0 result)
         and [l1] is smaller than [l2]
       - the empty list [[]] is strictly smaller than non-empty lists
   
       Note: the [cmp] function will be called even if the lists have
       different lengths.
   
       @since 4.12
   *)
   ```
   
   Good luck with that one as well.

For all the examples I have shown above, there are people who would easily grasp what the function does with a one-line code example, and will have trouble understanding the documentation as it is written. (By "grasp" I don't mean that they would be aware of all corner cases etc., but this can be easier to understand in the text than the actual description of what the function does because it is more concrete. Even the abstract description will be easier for them to attack once they already built a mental model of what the function does from the example.)

(For people who have experience doing and reading mathematical proofs, I think there are similarities: it can be very hard to follow someone else's precise proof, sometimes it is much easier to read a vague proof sketch or work out an example, build your own proof argument in your head (it may be imprecise/wrong), and then read the written proof with this argument already in mind.)
